packages/flame/lib/src/geometry/ray2.dart

```diff
@@ -0,0 +1,106 @@
+import 'dart:math';
+
+import 'package:flame/geometry.dart';
+import 'package:meta/meta.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+/// A ray in the 2d plane.
+///
+/// The [direction] should be normalized.
+class Ray2 {
+  Ray2(this.origin, Vector2 direction) {
+    this.direction = direction;
+  }
+
+  Ray2.zero() : this(Vector2.zero(), Vector2(1, 0));
+
+  Vector2 origin;
+  final Vector2 _direction = Vector2.zero();
+  Vector2 get direction => _direction;
+  set direction(Vector2 direction) {
+    assert(
+      (direction.length2 - 1).abs() < 0.000001,
+      'direction must be normalized',
+    );
+    _direction.setFrom(direction);
+    directionInvX = 1 / direction.x;
+    directionInvY = 1 / direction.y;
+  }
+
+  // These are the inverse of the direction (the normal), they are used to avoid
+  // a division in [intersectsWithAabb2], since a ray will usually be tried
+  // against many bounding boxes it's good to pre-calculate it, which is done
+  // in the direction setter.
+  @visibleForTesting
+  late double directionInvX;
+  @visibleForTesting
+  late double directionInvY;
+
+  /// Whether the ray intersects the [box] or not.
+  ///
+  /// Caveat: Since this uses the Branchless Ray/Bounding box intersection
+  /// method by Tavian, rays that originate and follow an edge of the AABB will
+  /// be considered as non-intersecting.
+  /// https://tavianator.com/2011/ray_box.html
+  /// https://tavianator.com/2015/ray_box_nan.html
+  bool intersectsWithAabb2(Aabb2 box) {
+    final tx1 = (box.min.x - origin.x) * directionInvX;
+    final tx2 = (box.max.x - origin.x) * directionInvX;
+
+    final ty1 = (box.min.y - origin.y) * directionInvY;
+    final ty2 = (box.max.y - origin.y) * directionInvY;
+
+    final tMin = max(min(tx1, tx2), min(ty1, ty2));
+    final tMax = min(max(tx1, tx2), max(ty1, ty2));
+
+    return tMax > max(tMin, 0);
+  }
+
+  /// Gives the point at a certain length along the ray.
+  Vector2 point(double length, {Vector2? out}) {
+    return ((out?..setFrom(direction)) ?? direction.clone())
+      ..scale(length)
+      ..add(origin);
+  }
+
+  static final Vector2 _v1 = Vector2.zero();
+  static final Vector2 _v2 = Vector2.zero();
+  static final Vector2 _v3 = Vector2.zero();
+
+  /// Returns where (length wise) on the ray that the ray intersects the
+  /// [LineSegment] or null if there is no intersection.
+  ///
+  /// A ray that is parallel and overlapping with the [segment] is considered to
+  /// not intersect.
```

Comment:
I wonder if we can handle this edge case better? For example, adding `if (dot == 0) { ... }`.

Or was this a deliberate choice?

Reply:
That would be nice, what should we return in that case though, would `t1` still be a valid point on the line segment?

Reply:
Ah, obviously it can't return `t1` since that would be having a division by zero.
I'll try around a bit, if you have a good idea of how it can be done I'm all ears.

Reply:
What do you think about the solution tgat I pushed now?

Reply:
Hmm, maybe it makes sense that it doesn't collide with parallel rays, because it will always collide with one of the line segments connecting to it anyways?

Reply:
I guess it makes sense, esp since the intersection in this case is not a single point but rather a segment, so returning a single point would be wrong.
But we should probably explain this in a comment, because I suspect we'll soon forget what the reason was.

Reply:
Good idea, I added that in #1785